### ISSUE TITLE
Camera Preview Start Up Delay

### ISSUE THREAD

**Issue by jtkeyva** [fluttercandies/flutter_wechat_camera_picker#96]
**Describe the bug**
Camera preview takes 2x as long as regular camera package to show the preview.

**How to reproduce**
Call the CameraPicker.pickFromCamera function using any config.

Open the wechat camera picker and wait 2+ seconds of black to see actual camera preview

**Expected behavior**
The camera preview screen to show nearly instantaneously or under 1 second like the Camera package.
I noticed some apps grab 1 frame and sometimes blur it until the live preview fully kicks in. perhaps that's their workaround?

**Screenshots (If contains)**

**Version information**

- iPhone 12
- iOS 15.3
- Package Version: wechat_camera_picker 3.2.0+1
- Flutter Version: 2.10 and/or 3

**Additional context**
I don't doubt if there's an unavoidable delay before camera preview appears, but this package in particular takes extra long compared to other camera packages. it may be by design to fade in from black rather than an abruptly appearing video preview, but i think that should be an option.

Also, if there would be a way to keep the camera preview loaded for instant reveal, that would be useful. Or leave camera preview going for 1-2 seconds after it's use in case you want to flip back open the camera real quick.

**Comment by AlexV525** [comment]
This is intended. We have an existing issue targeting `camera` tracking this: https://github.com/flutter/flutter/issues/57831

TL;DR, we want to provide the best experience when recording without any black/blank frames. To do so, proper preparation must be called before recording:
https://github.com/fluttercandies/flutter_wechat_camera_picker/blob/2b29dfa0e6d0bb946beab00b263cb8809e39af90/lib/src/widgets/camera_picker.dart#L416-L418

It's a trade-off between missing frames and a longer initialization duration. No matter where you put the preparation, that would be unavoidable to wait. We've compared various experiences by putting the preparation into different callers, such as just before the recording started. The best would be at the initialization at present.

To avoid waiting when you don't want to record, you can disable the record and the preview should be the same compared to other camera packages. Otherwise, I would consider it as not doable.

Closing as it's a won't fix issue, but feel free to leave more thoughts.

**Comment by jtkeyva** [comment]
Yeah I hear you. I prefer a slight jank over just a black screen. Black feels like something is lagging or broken. 

How about an option to show a custom widget during initialization so we can add shimmer or color fading animation or circularprogressindicator or whatever. Similar to your overlay option but during the initialization period.

What I've done is immediately record a 300ms blip of video as soon as the camera initializes. There's a little hiccup/jank but then the recording is always smooth afterwards and no black frames. Assuming that's what you're doing behind the black?

I also like the idea/option/feature of showing the camera preview immediately, recording a 300ms and blitting it to the screen and possibly blurring it until the camera is fully ready. I like the fade in of the camera you have for sure, but plain black just feels like something is missing.

Also, wanted to note that this delay happens when switching cameras from front to back etc.

**Comment by AlexV525** [comment]
> Yeah I hear you. I prefer a slight jank over just a black screen. Black feels like something is lagging or broken.

It's not only a jank produced, but another black screen because it starts another capture session.

> How about an option to show a custom widget during initialization so we can add shimmer or color fading animation or circularprogressindicator or whatever. Similar to your overlay option but during the initialization period.

Could be a good option for the experience, I thought about a tip like "Waiting for the camera to be ready..." or some other things. We already have a foreground builder when the controller is initialized, maybe it's easier to implement this request by making the builder available when the controller is not ready.

> What I've done is immediately record a 300ms blip of video as soon as the camera initializes. There's a little hiccup/jank but then the recording is always smooth afterwards and no black frames. Assuming that's what you're doing behind the black?

Do you mean using other camera packages? If so, then yes.

> Also, wanted to note that this delay happens when switching cameras from front to back etc.

All controllers benefit from the same preparation so yes, they will have the same delay.

**Comment by jtkeyva** [comment]
Got it. Perhaps you could clone the foregroundBuilder and just make an initilization builder so people can show whatever they want during that brief period.

Great job on the package, lot's of nice details!

Btw, is there any way to minimize the camera to a little floating window? Or is that out of the scope of the project?

**Comment by AlexV525** [comment]
> Got it. Perhaps you could clone the foregroundBuilder and just make an initilization builder so people can show whatever they want during that brief period.

You can try with this https://github.com/fluttercandies/flutter_wechat_camera_picker/pull/97 to see if it's suitable for your case. The builder is valid when the `cameraValue` is `null` which indicates the controller is not yet initialized.

> Btw, is there any way to minimize the camera to a little floating window? Or is that out of the scope of the project?

This is indeed out of scope. However, I think it's doable by combining the `foregroundBuilder` with a custom page route, though it might have other UI issues with the picker layout when the `MediaQuery` is narrowed.

**Comment by jtkeyva** [comment]
Great job, working fine here. Thanks!

Is it up to the dev to program when to stop their CircularProgressIndicator/loading animation? Do you get any indicator once the camera is fully live and visible?

**Comment by AlexV525** [comment]
> Do you get any indicator once the camera is fully live and visible?

The package can only tell whether the controller is initialized, without the information whether the camera preview is visible.

**Comment by AlexV525** [comment]
For the record, this is how long the preparation takes:
![image](https://user-images.githubusercontent.com/15884415/172139420-82ea4a8b-a495-420a-9e4f-c8b009c308fa.png)

**Comment by jtkeyva** [comment]
Got it thanks! Any way to fire off a notification when "Prepared for video recording" hits the log? That way we don't have to muck with timers or guess anything. Or perhaps have an initializingCameraBuilder that shows until "Prepared for video recording" has fired.

**Comment by AlexV525** [comment]
> Any way to fire off a notification when "Prepared for video recording" hits the log? That way we don't have to muck with timers or guess anything. Or perhaps have an initializingCameraBuilder that shows until "Prepared for video recording" has fired.

Not sure what you need to guess here. The foreground builder is available for the initializations now. If you have more feature request, please file new issues.
